cmd/buckets-scan-mgr.go

```diff
@@ -0,0 +1,278 @@
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cmd
+
+import (
+	"context"
+	"sync"
+	"time"
+)
+
+/*
+
+bucketsScanMgr decides what is the next bucket to scan for a given pool/set. It
+holds internal information about when the last time a bucket was finished scanned
+to select buckets with the oldest last scan for the next scan operation.
+
+A minimal use of the bucketsScanMgr is by doing this:
+```
+  mgr := newBucketsScanMgr(ObjectLayer)
+  for bucket := range mgr.getBucketCh() {
+      ...
+  }
+```
+
+*/
+
+const (
+	// the interval to discover if there are new buckets created in the cluster
+	bucketsListInterval = time.Minute
+)
+
+type setID struct {
+	pool, set int
+}
+
+type bucketScanStat struct {
+	ongoing      bool      // this bucket is currently being scanned
+	lastFinished time.Time // the last cycle of this scan
+	lastUpdate   time.Time
+	cycle        uint32 // the last cycle of this scan
+	lifecycle    bool   // This bucket has lifecycle set
+}
+
+type bucketsScanMgr struct {
+	ctx context.Context
+
+	// A registered function which knows how to list buckets
+	bucketsLister func(context.Context, BucketOptions) ([]BucketInfo, error)
+
+	mu           sync.RWMutex
+	knownBuckets map[string]struct{}                 // Current buckets in the S3 namespace
+	bucketsCh    map[setID]chan string               // A map of an erasure set identifier and a channel of buckets to scan
+	internal     map[setID]map[string]bucketScanStat // A map of an erasure set identifier and bucket scan stats
+}
+
+func newBucketsScanMgr(s3 ObjectLayer) *bucketsScanMgr {
+	mgr := &bucketsScanMgr{
+		ctx:           GlobalContext,
+		bucketsLister: s3.ListBuckets,
+		internal:      make(map[setID]map[string]bucketScanStat),
+		bucketsCh:     make(map[setID]chan string),
+	}
+	return mgr
+}
+
+func (mgr *bucketsScanMgr) getKnownBuckets() []string {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+
+	ret := make([]string, 0, len(mgr.knownBuckets))
+	for k := range mgr.knownBuckets {
+		ret = append(ret, k)
+	}
+	return ret
+}
+
+func (mgr *bucketsScanMgr) isKnownBucket(bucket string) bool {
+	mgr.mu.Lock()
+	defer mgr.mu.Unlock()
+
+	_, ok := mgr.knownBuckets[bucket]
+	return ok
+}
+
+func (mgr *bucketsScanMgr) start() {
+	m := &sync.Mutex{}
+	c := sync.NewCond(m)
```

Comment:
I'd much prefer the simplicity of...
```suggestion
	m := make(chan struct{}, 1)
```

Then do a select with default to send. and a simple `<-m` to wait for the first. 

Technically what you have here is racy, since the broadcast can be done before you are waiting for it - and you'll miss it.